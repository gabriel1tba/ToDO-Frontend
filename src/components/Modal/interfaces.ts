export interface IModal {
  title: string;
  children: React.ReactNode;
  open: boolean;
  onCloseModal: () => void;
}
