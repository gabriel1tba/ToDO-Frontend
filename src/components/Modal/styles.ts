import styled from 'styled-components';

export const Overlay = styled.div`
  z-index: 1;
  position: fixed;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  display: flex;
  align-items: center;
  justify-content: center;
  background: rgba(0, 0, 0, 0.6);
`;

export const Wrapper = styled.div`
  z-index: 2;

  position: relative;
  text-align: flex;
  width: 550px;
  background: #fff;
  border-radius: 5px;
  border: 1px solid #ced4da;
  header {
    margin: 15px 0;
    border-bottom: 1px solid #ced4da;

    > p {
      font-size: 20px;
      color: #343a40;

      margin: 0 0 15px 21px;

      cursor: default;
    }

    button {
      position: absolute;
      top: 15px;
      right: 20px;
      background: transparent;
      border: none;

      > svg {
        cursor: pointer;

        transition: transform 0.3s;

        &:hover {
          transform: scale(1.2);
        }
      }
    }
  }
`;

export const ModalWrapper = styled.div`
  margin: 40px 20px;
`;
