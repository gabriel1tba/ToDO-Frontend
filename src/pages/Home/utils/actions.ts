import { TTodo } from 'services/TodoService/interfaces';

export enum Types {
  GET_TODOS = 'GET_TODOS',
  CREATE_TODO = 'CREATE_TODO',
  UPDATE_TODO = 'UPDATE_TODO',
  DELETE_TODO = 'DELETE_TODO',
}

interface GetTodos {
  type: Types.GET_TODOS;
  payload: TTodo[];
}

interface CreateTodo {
  type: Types.CREATE_TODO;
  payload: TTodo;
}

interface UpdateTodo {
  type: Types.UPDATE_TODO;
  payload: TTodo;
}

interface DeleteTodo {
  type: Types.DELETE_TODO;
  payload: string;
}

export type Action = GetTodos | CreateTodo | UpdateTodo | DeleteTodo;

export const getTodosAction = (todos: TTodo[]): GetTodos => ({
  type: Types.GET_TODOS,
  payload: todos,
});

export const createTodoAction = (todo: TTodo): CreateTodo => ({
  type: Types.CREATE_TODO,
  payload: todo,
});

export const updateTodoAction = (todo: TTodo): UpdateTodo => ({
  type: Types.UPDATE_TODO,
  payload: todo,
});

export const deleteTodoAction = (id: string): DeleteTodo => ({
  type: Types.DELETE_TODO,
  payload: id,
});
