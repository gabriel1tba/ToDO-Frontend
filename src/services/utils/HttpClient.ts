import axios from 'axios';

const HttpClient = axios.create({
  baseURL: `${process.env.REACT_APP_API_HOST}:${process.env.REACT_APP_API_PORT}`,
});

export default HttpClient;
