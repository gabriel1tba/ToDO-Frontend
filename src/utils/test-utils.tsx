import { render, RenderOptions } from '@testing-library/react';
import { TodoProvider } from 'context/todos';

interface IWrapperProvider {
  children: React.ReactNode;
}

const WrapperProvider = ({ children }: IWrapperProvider) => (
  <TodoProvider>{children}</TodoProvider>
);

const customRender = (ui: React.ReactElement, options?: RenderOptions) =>
  render(<WrapperProvider>{ui}</WrapperProvider>, { ...options });

export * from '@testing-library/react';
export { customRender as render };
