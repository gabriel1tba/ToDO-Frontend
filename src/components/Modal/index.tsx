import * as S from './styles';

interface IModal {
  handleCloseModal: () => void;
}

const Modal = ({ handleCloseModal }: IModal) => {
  return (
    <S.Overlay>
      <S.Wrapper>
        <header>Editar</header>

        <strong>Parabéns</strong>
        <p>Você alcançou um novo level</p>

        <button onClick={handleCloseModal}>
          <img src="/icons/close.svg" alt="Icone de fechar" />
        </button>
      </S.Wrapper>
    </S.Overlay>
  );
};

export default Modal;
