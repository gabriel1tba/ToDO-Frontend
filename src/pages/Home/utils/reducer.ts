import { TTodo } from 'services/TodoService/interfaces';

import { Types, Action } from './actions';

const todosReducer = (state: TTodo[], action: Action): TTodo[] => {
  switch (action.type) {
    case Types.GET_TODOS: {
      return action.payload;
    }

    case Types.CREATE_TODO: {
      return [...state, action.payload];
    }

    case Types.UPDATE_TODO: {
      return state.map((todo) => {
        if (action.payload.id === todo.id) {
          todo = { ...action.payload, created_at: todo.created_at };
        }
        return todo;
      });
    }

    case Types.DELETE_TODO: {
      return state.filter((todo) => {
        return todo.id !== action.payload;
      });
    }

    default:
      return state;
  }
};

export default todosReducer;
