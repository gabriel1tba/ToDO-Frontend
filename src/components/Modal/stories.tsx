import { Story, Meta } from '@storybook/react';

import Modal, { IModal } from '.';

export default {
  title: 'Modal',
  component: Modal,
  argTypes: {
    children: {
      type: 'string',
    },
    title: {
      type: 'string',
    },
    open: {
      type: 'boolean',
    },
    onCloseModal: {
      type: 'function',
    },
  },
} as Meta;

export const Default: Story<IModal> = (args) => <Modal {...args} />;

Default.args = {
  children: 'Content in modal',
  title: 'Content On Modal',
  open: true,
};
