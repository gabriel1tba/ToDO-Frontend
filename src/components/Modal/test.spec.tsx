import { render, screen, waitFor } from 'utils/test-utils';
import userEvent from '@testing-library/user-event';

import Modal from '.';

describe('<Modal />', () => {
  beforeEach(() => {
    render(<div id="portal-modal-root" />);
  });

  it('should render the Modal correctly', () => {
    const { container } = render(
      <Modal title="Cadastro de pessoas" open onCloseModal={() => ({})}>
        <form>
          <input type="text" placeholder="Digite seu nome" />
          <input type="email" placeholder="Digite seu email" />

          <button>Fechar</button>
          <button>Confirmar</button>
        </form>
      </Modal>,
    );

    expect(screen.getByText(/cadastro de pessoas/i)).toBeInTheDocument();

    expect(screen.getAllByRole('button')).toHaveLength(3);

    expect(screen.getAllByRole('textbox')).toHaveLength(2);

    expect(screen.getByPlaceholderText(/digite seu nome/i)).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText(/digite seu email/i),
    ).toBeInTheDocument();

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should no render Modal if open is false', () => {
    render(
      <Modal title="Cadastro de pessoas" open={false} onCloseModal={() => ({})}>
        <h1>Content Modal</h1>
      </Modal>,
    );

    expect(screen.queryByText(/content modal/i)).not.toBeInTheDocument();
  });

  it('should the method onCloseModal must be called correctly', async () => {
    const onCloseModal = jest.fn();

    render(
      <Modal title="Cadastro de pessoas" open onCloseModal={onCloseModal}>
        <h1>Content Modal</h1>
      </Modal>,
    );

    userEvent.click(screen.getAllByRole('button')[0]);

    await waitFor(() => {
      expect(onCloseModal).toHaveBeenCalled();
    });
  });
});
