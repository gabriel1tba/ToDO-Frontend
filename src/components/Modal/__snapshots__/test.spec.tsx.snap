// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should render the Modal correctly 1`] = `
.c0 {
  position: absolute;
  right: 0;
  top: 0;
  padding: 30px;
  overflow: hidden;
  z-index: 3;
}

<div
  class="c0"
/>
`;
